### PR TITLE
[JAX] flash attention integration

### PR DESCRIPTION
**Pull request by zlsh80826**
This PR exposes the cuDNN arbitrary seqlen fused attention backend to JAX.

Main changes:
- [x] Support `seqlen%64==0` and `head_dim=128` self attention
- [x] Remove `_max_512_` suffix in the related functions
- [x] Adding `output`, `rng_state` as the ctx tensors since flash attention requires them
- [x] Allocate different softmax ctx tensors size for different backends
- [x] Set different increment of offset for different backends
- [x] Add long sequence length fused attention unit tests
- [x] `jax.abstract_arrays.Shaped` is deprecated. Replace it with `jax.core.ShapedArray`

### CODE REVIEW

**Comment by jeng1220** [comment]
All CI tests were passed.
@timmoon10 and @cyanguwa please review the commits, and merge the code if everything looks good.

**Review by timmoon10 — APPROVED** [review]
LGTM

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -137,82 +195,27 @@ def set_input(self, b, s, h, d, *, attn_bias_type, attn_mask_type, dropout_proba
...
-                       s,
-                       h,
-                       d,
-                       attn_bias_type=attn_bias_type,
-                       attn_mask_type=attn_mask_type,
-                       dropout_probability=dropout_probability,
-                       dtype=dtype,
-                       is_training=is_training,
-                       pad_ratio=pad_ratio)
+    def test_forward(self, b, s, h, d, attn_bias_type, attn_mask_type, backend, dropout_probability,
```
Comment (cyanguwa) [inline tests/jax/test_fused_attn.py:205]: If we have `test_forward_backward`, maybe we don't need `test_forward` anymore? Or am I misreading the code?

File: transformer_engine/jax/csrc/modules.cpp [hunk 2]
```diff
@@ -768,78 +784,81 @@ void SelfFusedAttnMax512Forward(cudaStream_t stream, void **buffers, const char
...
+    auto s_tensor = TensorWrapper(nullptr, std::vector<size_t>{1}, dtype);
 
-    auto dummy_rng_state_tensor = TensorWrapper(nullptr, std::vector<size_t>{2}, DType::kInt64);
+    // aux tensors
+    auto rng_state_tensor = TensorWrapper(rng_state, std::vector<size_t>{2}, DType::kInt64);
+
+    auto backend = nvte_get_fused_attn_backend(
+        static_cast<NVTEDType>(dtype), static_cast<NVTEDType>(dtype), qkv_layout, bias_type,
+        mask_type, dropout_probability, q_max_seqlen, kv_max_seqlen, head_dim);
+    PopulateRngStateAsync(rng_state, seed, q_max_seqlen, kv_max_seqlen, backend, stream);
```
Comment (cyanguwa) [inline transformer_engine/jax/csrc/modules.cpp:806]: Maybe rename this to something like PopulateInputParamsAsync? Nitpick.

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:205]: I think both forward and backward needs their own unit tests so that if one of them is failed we can quickly know which fails. There is no standalone `test_backward` because the backward needs the forward output.

Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:205]: Another reason is that we uses `value_and_grad` in the `test_forward_backward` for verifying the [autodiff](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#evaluate-a-function-and-its-gradient-using-value-and-grad), which requires forward output as a scalar (like loss). So we can't verify the forward output elementwise, therefore, a standalone forward is necessary.

File: transformer_engine/jax/csrc/modules.cpp (hunk 2 above)
Comment (zlsh80826) [reply transformer_engine/jax/csrc/modules.cpp:806]: The [`PopulateRngStateAsync`](https://github.com/NVIDIA/TransformerEngine/blob/3e791357921f4f86c7925011909f60ffbfeebd8f/transformer_engine/jax/csrc/utils.cu#L43) literally populates the rng_state only, so I preferred the original name.

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:205]: I hanced the docstring of the unit test in https://github.com/NVIDIA/TransformerEngine/pull/345/commits/8d8ff5d9fdfd2e5eb12ca9ec778cc26bf0632c9e. Hope it is more descriptive.
